source/geod24/concurrency.d

```diff
@@ -1643,14 +1692,56 @@ public class Channel (T)
     public bool send (T msg)
     {
         this.mutex.lock();
-        scope (exit)
-            this.mutex.unlock();
 
         if (this.closed)
+        {
+            this.mutex.unlock();
             return false;
+        }
+
+        if (!this.recvq.empty)
+        {
+            ChannelContext!T context = this.recvq.front;
+            this.recvq.removeFront();
+            *(context.msg_ptr) = msg;
+            this.mutex.unlock();
+            context.notify();
+            return true;
+        }
+
+        if (this.mcount < this.qsize)
+        {
+            this.queue.insertBack(msg);
+            this.mcount++;
+            this.mutex.unlock();
+            return true;
+        }
 
-        this.queue.insertBack(msg);
-        return true;
+        scope scheduler = thisScheduler;
```

Comment:
```suggestion
        auto scheduler = thisScheduler;
```

Just a note here: `scope` has no meaning when you assign existing values to variables. It only has meaning when you instantiate a new object. For example:

```D
scope obj = new Class();  // destroyed at end of scope
```

This is different to the following:

```D
auto obj = new Class();
scope var = obj;  // *not* destroyed at end of scope
```